### PR TITLE
[flang][OpenMP] fix lastprivate for allocatables

### PR DESCRIPTION
**Pull request by tblah**
Don't use `copyHostAssociateVar` for allocatable variables. It isn't
clear to me whether or not this should be addressed in
`copyHostAssociateVar` instead of inside OpenMP. I opted for OpenMP
to minimise how many things I effected. `copyHostAssociateVar` will
not update the destination variable if the destination variable
was unallocated. This is incorrect because assignment inside of the
openmp block can cause the allocation status of the variable to
change. Furthermore, `copyHostAssociateVar` seems to only copy the
variable address not other metadata like the size of the allocation.
Reallocation by assignment could cause this to change.

Non-OpenMP reviewers: should this actually be done inside of `copyHostAssociateVar`?

OpenMP reviewers: I am aware that this will change with delayed privatization. My intention here is to squash the bug before the LLVM branch

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
> Non-OpenMP reviewers: should this actually be done inside of copyHostAssociateVar?

It need not be. copyHostAssociateVar has a few customers and only OpenMP uses it for lastprivatisation. Previously the symbol table was not available in OpenMP.cpp. So this could only be done in the bridge. With changes to support recursive lowering, OpenMP.cpp has the Symbol table available.

**Comment by kiranchandramohan** [comment]
> OpenMP reviewers: I am aware that this will change with delayed privatization. My intention here is to squash the bug before the LLVM branch

At the moment delayed privatisation has no support for lastprivate. When we add support for it then this new change should be usable.


> realloc_lhs
In later work, we should investigate whether this is needed.

**Comment by tblah** [comment]
> > realloc_lhs
> 
> In later work, we should investigate whether this is needed.

I presume you mean `temporary_lhs`. This is tricky - the `lhs` of the assignment is not a temporary so it is a missuse of the flag. I originally included it because the documentation for `hlfir.assign` says that it prevents finalization of the lhs or lhs subobjects. However, when I tried more testcases it turned out this prevented arrays of different lengths from being assigned. This seems to contradict the documentation for `realloc`, but I wanted to get this landed before the branch if possible.

Now that I am looking closer, it isn't clear to me whether calling the finalization would actually be a bug. The OpenMP 5.2 standard (section 5.4.5) says the update should occur as though by intrinsic assignment (which I believe would run finalization). But I might have missed something, because the finalization would have already run inside each thread when the private copy was overwritten.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 1]
```diff
@@ -127,8 +128,52 @@ void DataSharingProcessor::copyFirstPrivateSymbol(
...
+      builder.restoreInsertionPoint(*lastPrivIP);
+    else
+      builder.setInsertionPointAfter(sb.getAddr().getDefiningOp());
+
+    hlfir::Entity dst{hsb.getAddr()};
+    hlfir::Entity src{sb.getAddr()};
+    builder.create<hlfir::AssignOp>(
+        loc, src, dst, /*isWholeAllocatableAssignment=*/allocatable,
+        /*keepLhsLengthInAllocatableAssignment=*/false,
+        /*temporary_lhs=*/false);
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:170]: If this code is placed in the `isAllocatable` part of `copyVarHLFIR` in Bridge.cpp, will there be lot of tests that have to be updated?

Can we do this at some point in the future after the branch?

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Review by jeanPerier — APPROVED** [review]
Thanks Tom. Moving this into the copyHostAssociateVar/copyVarHLFIR would make sense to me.

I updated the allocatable handling part that was broken a while ago with the assumption that reallocation was not a thing inside the OpenMP region (https://github.com/llvm/llvm-project/pull/69441). If it is, then what you are doing is the right approach..

**Comment by tblah** [comment]
> Thanks Tom. Moving this into the copyHostAssociateVar/copyVarHLFIR would make sense to me.
> 
> I updated the allocatable handling part that was broken a while ago with the assumption that reallocation was not a thing inside the OpenMP region (#69441). If it is, then what you are doing is the right approach..

Thanks for taking a look. Explicit reallocation is not allowed inside of an OpenMP region but I think allocation is allowed implicitly via intrinsic assignment or privatisation (which works the same as intrinsic assignment).

One of our internal tests failed when I tried to put this in copyHostAssociateVar. I will investigate this and post a followup patch if I get it working.

**Pull request merged**